From eadef8ac233fe64b00d42d1c259552e74ae0a467 Mon Sep 17 00:00:00 2001
From: Ashok Raj Deenadayalan <adeena@codeaurora.org>
Date: Wed, 21 Jan 2015 12:53:28 +0530
Subject: [PATCH 176/176] Camera: Fix Flash issue upon exit from Camcorder

Upon exit from Camcorder with FLASH ON, the FLASH LED
continues to be ON for around 8 sec. Force turn-off
flash upon Camera exit to fix this issue.

Change-Id: Icf8b51a6cf583ae6ac05b309060486a25d91635a
CRs-Fixed: 777634
---
 QCamera/stack/mm-camera-interface/inc/mm_camera.h |  1 +
 QCamera/stack/mm-camera-interface/src/mm_camera.c | 14 ++++++++++++++
 2 files changed, 15 insertions(+)

diff --git a/QCamera/stack/mm-camera-interface/inc/mm_camera.h b/QCamera/stack/mm-camera-interface/inc/mm_camera.h
index e96e8b6..dbd0a69 100755
--- a/QCamera/stack/mm-camera-interface/inc/mm_camera.h
+++ b/QCamera/stack/mm-camera-interface/inc/mm_camera.h
@@ -450,6 +450,7 @@ typedef struct mm_camera_obj {
     uint32_t snap_burst_num_by_user;
     camera_mode_t current_mode;
     uint32_t op_mode;
+    uint8_t flash_disabled;
     cam_ctrl_dimension_t dim;
 } mm_camera_obj_t;
 
diff --git a/QCamera/stack/mm-camera-interface/src/mm_camera.c b/QCamera/stack/mm-camera-interface/src/mm_camera.c
index 42b9215..06cab76 100755
--- a/QCamera/stack/mm-camera-interface/src/mm_camera.c
+++ b/QCamera/stack/mm-camera-interface/src/mm_camera.c
@@ -331,6 +331,8 @@ int32_t mm_camera_open(mm_camera_obj_t *my_obj)
     /* set geo mode to 2D by default */
     my_obj->current_mode = CAMERA_MODE_2D;
 
+    my_obj->flash_disabled = 1;
+
     pthread_mutex_init(&my_obj->cb_lock, NULL);
 
     CDBG("%s : Launch async cmd Thread in Cam Open",__func__);
@@ -371,6 +373,17 @@ on_error:
 
 int32_t mm_camera_close(mm_camera_obj_t *my_obj)
 {
+    int32_t value = 0; /* Value to turn off LED */
+    /* Flash is left ON in few use case like Recording with flash ON.
+       Turn it off here */
+
+    if (!my_obj->flash_disabled) {
+        mm_camera_send_native_ctrl_cmd(my_obj,
+                                       CAMERA_SET_PARM_LED_MODE,
+                                       sizeof(int32_t),
+                                       value);
+    }
+
     CDBG("%s : Close evt Poll Thread in Cam Close",__func__);
     mm_camera_poll_thread_release(&my_obj->evt_poll_thread);
 
@@ -1527,6 +1540,7 @@ int32_t mm_camera_set_general_parm(mm_camera_obj_t * my_obj,
                                             CAMERA_SET_PARM_LED_MODE,
                                             sizeof(int32_t),
                                             p_value);
+        my_obj->flash_disabled = !(*(uint8_t*)p_value);
         break;
     case MM_CAMERA_PARM_ROLLOFF:
         rc = mm_camera_send_native_ctrl_cmd(my_obj,
-- 
1.8.3.1

